engine.fhir/src/main/java/org/opencds/cqf/cql/engine/fhir/model/Dstu2FhirModelResolver.java

```diff
@@ -47,7 +47,7 @@ public Dstu2FhirModelResolver() {
 		this(FhirContext.forDstu2());
 	}
 
-	private Dstu2FhirModelResolver(FhirContext fhirContext) {
+    public Dstu2FhirModelResolver(FhirContext fhirContext) {
```

Comment:
These need to remain private. Due to the way FhirContexts are initialized it's not possible to pass in any ole FhirContext and have it work.

Reply:
See this issue for further discussion:
https://github.com/DBCG/cql_engine/issues/360

Reply:
Consider the case that an external user has registered their own custom types. They may have subclassed something like an `Encounter` in such a way as breaks the behavior expected by the CQL engine. Even calling the `scanResourceTypes` function directly wouldn't override that subclassed implementation due to the way the maps are stored.

Reply:
You'd have to effectively replace all the custom implementations with the generic HAPI ones to ensure the behavior remained the same. Doing that would invalidate the `FhirContext` usage outside of the `ModelResolver`.

Reply:
In other words, based on experience, the HAPI contract defined by the API isn't sufficiently strong to prevent breakage in the cql-engine when using custom types.

Reply:
If we very clearly documented the expectations and/or added checks such that we ensured the FhirContexts were in a usable state (such as not yet initialized or initialized with all the necessary default types) I could definitely be persuaded to make these public since it's a performance win. I don't know how to do that generally though. If an external user registered only `Encounter` and `Patient` because that's all their FHIR Server needed, but then we executed some QI Core CQL that required a broader set of `Resource`s, it's a nasty runtime error. IOW, open to suggestions, but it's been a pain point in the past.

Reply:
But that should be allowed, no? 

If a dev overrides `Encounter` and custom loads it into their already loaded FhirContext AND proceeds to pass that same context to the Engine's loader (they don't need to, they can create a new Context in the default constructor), I would assume the developer wants to use that custom class inside the engine as well.

Reply:
> If an external user registered only `Encounter` and `Patient` because that's all their FHIR Server needed, but then we executed some QI Core CQL that required a broader set of `Resource`s, it's a nasty runtime error. 

Hum... this should not happen anymore (I remember testing this when doing that new reflection method). Would an additional test case with an incomplete FhirContext suffice?

Reply:
There's no documentation that they have to extend _both_ the `ModelResolver` _and_ any custom types to make this work. This absolutely does fail in unexpected ways for new users of the API. Part of the reason the `setPackageNames` API is now public was to make the cql_engine work in non-HAPI environments like the LinuxForHealth FHIR server. This API used to be public and allow passing in any FhirContext. It caused a ton of support requests for very subtle issues. I'm not talking hypotheticals here. This was specifically taken from being public to being private because of the large number of support issues it caused:

https://github.com/DBCG/cql_engine/commit/d3059050f64d4f9525b6bf25db4605de6833fe5f

Reply:
Hum.. In the past, did people had issues because the code would not initialize or because they would change the context so much that would break the engine? 

The previous code did allow many initialization issues. It was super easy to happen. Virtually any semi-loaded context would indeed crash. I am assuming that was the main problem. The current one is supposed to work no matter what you through at it. You have to really break the compatibility of the classes to crash it.

Reply:
This code works well if you're not using any extended classes, but that's actually not necessarily the most common scenario in the deployments we support. We have several using custom types.

There's also the issue that the code isn't thread-safe as written. The FhirContext may be concurrently modified by some external process, and the type-registration code is explicitly not thread-safe:
https://github.com/hapifhir/hapi-fhir/blob/master/hapi-fhir-base/src/main/java/ca/uhn/fhir/context/FhirContext.java#L967

If you have a shared `FhirContext` on CQL execution server serving many different requests with one re-used context, what happens if the CQL engine tries to initialize the context at the same time as some other non-CQL operation?

Reply:
> This code works well if you're not using any extended classes, but that's actually not necessarily the most common scenario in the deployments we support. We have several using custom types.

Then just don't reuse the Context. We are not forcing people to use the cached version. In fact, the default is a brand new context. It feels silly to block all the other simpler use cases because the devs behind complicated ones MIGHT misuse the library. 

> There's also the issue that the code isn't thread-safe as written.

Right now the recommendation is to create the Resolver and get the FhirContext from inside of it to use it externally. If you are saying that the FhirContext should be fully encapsulated by each new Resolver in order to be thread-safe, then we need to change the current implementation and recommendation. 

> If you have a shared FhirContext on CQL execution server serving many different requests with one re-used context

Again, this change is not forcing the re-use. If devs want to try to share the context among everyone why shouldn't they be able to? 

FYI: Most simple implementations will end up extending each modelResolver just to call the initialize method with a cached Context anyway. I already do this with the [WHO Verifier](https://github.com/WorldHealthOrganization/ddcc-validator/blob/f373a6fc2a1af7abfb8bfaf5f378d25e9eb11f8f/app/src/main/java/org/who/ddccverifier/services/cql/CQLEvaluator.kt#L17) and the SDK would have to do the same.

Reply:
> Then just don't reuse the Context

This is the safe default for all the reasons listed above. Again, this is not academic. I had a flood of support requests when it was the other way around. I'm not saying this _might_ happen. I'm saying I spent hours across several different implementations debugging race conditions and initialization issues. Mostly on a volunteer basis. This change _will_ create lots of problems for me. That's why I'm arguing against it despite the obvious performance benefits.

> we need to change the current implementation and recommendation.

The current implementation initializes the context before sharing, thereby bypassing the threading issue. Once the context is initialized, it's safe to share. So this point is not relevant.

> If devs want to try to share the context among everyone why shouldn't they be able to?

We could do this by making the constructors `protected`. Then implementors who understood the requirements could share as they saw fit.

Reply:
`protected` it is then!

Reply:
I made it protected, created a subclass for each Resolver in the test folder and adapted the test classes to use the Cached Version. Let me know if you like this approach.